src/Routing/Url.php

```diff
@@ -0,0 +1,272 @@
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.5.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Routing;
+
+use ArrayAccess;
+use InvalidArgumentException;
+
+/**
+ * URL Object
+ *
+ * Use this object whenever you need to generate an URL for your app.
+ *
+ * This is not thought for external URLs.
+ */
+class Url implements ArrayAccess
+{
+
+    /**
+     * Array representation of the URL
+     *
+     * @var array
+     */
+    protected $url = [];
+
+    /**
+     * Absolute URL
+     *
+     * @var bool
+     */
+    protected $absolute;
+
+    /**
+     * Sets the action
+     *
+     * @param string $action Controller action name
+     * @return $this
+     */
+    public function setAction($action)
+    {
+        $this->url['action'] = $action;
+
+        return $this;
+    }
+
+    /**
+     * Sets the controller
+     *
+     * @param string $controller Controller class name without the controller suffix
+     * @return $this
+     */
+    public function setController($controller)
+    {
+        $this->url['controller'] = $controller;
+
+        return $this;
+    }
+
+    /**
+     * Set prefix
+     *
+     * @param string $prefix Prefix
+     * @return $this
+     */
+    public function setPrefix($prefix)
+    {
+        $this->url['prefix'] = $prefix;
+
+        return $this;
+    }
+
+    /**
+     * Sets the plugin
+     *
+     * @param string|null $plugin Plugin name
+     * @return $this
+     */
+    public function setPlugin($plugin)
+    {
+        $this->url['plugin'] = $plugin;
+
+        return $this;
+    }
+
+    /**
+     * Sets values to be passed in the URL
+     *
+     * @param string $pass Args to be passed to a controller action
+     * @return $this
+     */
+    public function addPass($pass)
+    {
+        $this->url[] = $pass;
+
+        return $this;
+    }
+
+    /**
+     * Sets a single custom route parameter
+     *
+     * @param string $key Key
+     * @param string|int|float $value Value
+     * @return $this
+     */
+    public function setParam($key, $value)
+    {
+        if (in_array($key, ['controller', 'action', 'plugin', 'prefix'])) {
+            throw new InvalidArgumentException('Use the according setter method instead.');
+        }
+
+        $this->url[$key] = $value;
+
+        return $this;
+    }
+
+    /**
+     * Set custom route parameters as key value pair
+     *
+     * @param array $params Array of key value parameters
+     * @return $this
+     */
+    public function setParams(array $params)
+    {
+        foreach ($params as $key => $value) {
+            $this->setParam($key, $value);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Sets a query param
```

Comment:
param => parameter. Can we generally write out such wods? I guess that wouldn't hurt.